elastic/apm-server
# add exponential backoff to es clients

## Motivation/summary

This is being spun out of work done in https://github.com/elastic/apm-server/pull/4981, with original motivation coming from https://github.com/elastic/apm-server/issues/4138.

After investigation, our non-libbeat esclients weren't respecting backoff and retry configuration set in `apm-server.yml`. This has been added, and a basic exponential backoff. The included library is [used as an example](https://github.com/elastic/go-elasticsearch/blob/4b40206692088570801280584e614027e6ce818b/_examples/bulk/indexer.go#L86) in `go-elasticsearch`, so it seemed like a decent place to start. I'm not sure what the policy is on bringing in external libraries though, and it's easy enough to create our own (it does have an MIT license, though).

Regarding testing, I just wanted to check that we're hooking up the retry function correctly, and that we're setting the retry count correctly. Any suggestions for simplifications or additions are welcome!

## Checklist




For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

1. run a web server that will fail (something like [this](https://gist.github.com/stuartnelson3/ddbf40dea8c49f0d0699f1ac481ac2ac))
2. build and run a command that uses an es client, like `./apm-server apikey create -E output.elasticsearch.backoff.init=3s`

output from the webserver should show the requests are spaced out.

## Related issues
